feg/gateway/services/n7_n40_proxy/n7/n7.go

```diff
@@ -39,3 +50,509 @@ func NewN7Client(cfg *PCFConfig) (*n7_sbi.ClientWithResponses, error) {
+	isNegative := false
+	if offsetMin < 0 {
+		isNegative = true
+		offsetMin = -offsetMin
+	}
+	tz := fmt.Sprintf("%02d:%02d", offsetMin/60, offsetMin%60)
+	if isNegative {
+		tz = "-" + tz
+	} else {
+		tz = "+" + tz
+	}
+	tzStr := sbi.TimeZone(tz)
+	return &tzStr
+}
+
+func getSbiPduSesionType(pduSessionType protos.PduSessionType) sbi.PduSessionType {
+	switch pduSessionType {
+	case protos.PduSessionType_IPV4:
+		return sbi.PduSessionTypeIPV4
+	case protos.PduSessionType_IPV4IPV6:
+		return sbi.PduSessionTypeIPV4V6
+	case protos.PduSessionType_IPV6:
+		return sbi.PduSessionTypeIPV6
+	case protos.PduSessionType_UNSTRUCTURED:
+		return sbi.PduSessionTypeUNSTRUCTURED
+	default:
+		return sbi.PduSessionTypeIPV4
+	}
+}
+
+func getSbiBool(val *bool) bool {
+	if val == nil {
+		return false
+	}
+	return *val
+}
+
+func GetPccRules(
+	policyDecision *n7_sbi.SmPolicyDecision,
+) ([]*protos.StaticRuleInstall, []*protos.DynamicRuleInstall, []string, []string) {
+	staticRules := []*protos.StaticRuleInstall{}
+	dynamicRules := []*protos.DynamicRuleInstall{}
+	rulesToRemove := []string{}
+	umIds := []string{}
+	if policyDecision.PccRules == nil {
+		return staticRules, dynamicRules, rulesToRemove, umIds
+	}
+
+	for ruleId, pccRule := range policyDecision.PccRules.AdditionalProperties {
+		if pccRule.PccRuleId == "" {
+			// This rule to be removed
+			rulesToRemove = append(rulesToRemove, ruleId)
+			continue
+		}
+		var activationTime, deactivationTime *timestamp.Timestamp
+		if pccRule.RefCondData != nil && *pccRule.RefCondData != "" {
+			condData, found := policyDecision.Conds.Get(*pccRule.RefCondData)
+			if found {
+				activationTime = ConvertToProtoTimeStamp(condData.ActivationTime)
+				deactivationTime = ConvertToProtoTimeStamp(condData.DeactivationTime)
+			}
+		}
+		policyRule := GetPolicyRuleProto(policyDecision, &pccRule)
+		if policyRule == nil {
+			// This is a static rule. Static rule only has pccRuleId and RefCondData
+			staticRule := &protos.StaticRuleInstall{
+				RuleId:           ruleId,
+				ActivationTime:   activationTime,
+				DeactivationTime: deactivationTime,
+			}
+			staticRules = append(staticRules, staticRule)
+			continue
+		}
+
+		dynamicRule := &protos.DynamicRuleInstall{
+			PolicyRule:       policyRule,
+			ActivationTime:   activationTime,
+			DeactivationTime: deactivationTime,
+		}
+		dynamicRules = append(dynamicRules, dynamicRule)
+
+		if len(policyRule.MonitoringKey) > 0 {
+			umIds = append(umIds, string(policyRule.MonitoringKey))
+		}
+	}
+
+	return staticRules, dynamicRules, rulesToRemove, umIds
+}
+
+func GetPolicyRuleProto(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.PolicyRule {
+	policyRule := &protos.PolicyRule{
+		Id: pccRule.PccRuleId,
+	}
+	if pccRule.Precedence != nil {
+		policyRule.Priority = uint32(*pccRule.Precedence)
+	}
+	// For all Ref*Data Maximum one entry is expected.
+	// These are declared as arrays for future compatibility
```

Comment:
create functions to encapsulate all this logic as much as you can.
For example `func (pc *n7_sbi.PccRule)getRefUmData....`

You should not have long functions if this can be encapsulated on some kind of function with meaning